### PR TITLE
Add a new analyzer to detect code that can be updated to use object initializers.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
tagging @dotnet/roslyn-ide

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/CrefCompletionProviderTests.cs [hunk 1]
```diff
@@ -418,609 +418,5 @@ class C
 ";
             await VerifyProviderCommitAsync(text, "foo(int)", expected, '(', "foo");
         }
-
-        [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
-        public async Task CrefCompletionSpeculatesOutsideTrivia()
```
Comment (jmarolf) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/CrefCompletionProviderTests.cs]: >CrefCompletionSpeculatesOutsideTrivia [](start = 26, length = 37)

How is this related to this change?

**Comment by jmarolf** [comment]
> ' Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

While a good change, shouldn't this go in a separate PR?

---
Refers to: src/EditorFeatures/VisualBasicTest/PopulateSwitch/PopulateSwitchTests.vb:1 in ebfd1a9. [](commit_id = ebfd1a9970c87239be780d187178702335ca08de, deletion_comment = False)

File: src/Workspaces/Core/Portable/Editing/SyntaxGenerator.cs [hunk 2]
```diff
@@ -118,6 +118,10 @@ public SyntaxNode FieldDeclaration(IFieldSymbol field, SyntaxNode initializer)
                 initializer);
         }
 
+        //internal abstract SyntaxNode ObjectMemberInitializer(IEnumerable<SyntaxNode> fieldInitializers);
+        //internal abstract SyntaxNode NamedFieldInitializer(SyntaxNode name, SyntaxNode value);
+        //internal abstract SyntaxNode WithObjectCreationInitializer(SyntaxNode objectCreationExpression, SyntaxNode initializer);
```
Comment (jmarolf) [inline src/Workspaces/Core/Portable/Editing/SyntaxGenerator.cs]: remove?

File: src/Features/Core/Portable/GenerateMember/AbstractGenerateMemberService.cs [hunk 3]
```diff
@@ -90,7 +90,7 @@ protected AbstractGenerateMemberService()
 
             var syntaxFacts = document.Project.LanguageServices.GetService<ISyntaxFactsService>();
             var semanticModel = document.SemanticModel;
-            var isMemberAccessExpression = syntaxFacts.IsMemberAccessExpression(simpleNameOrMemberAccessExpression);
+            var isMemberAccessExpression = syntaxFacts.IsSimpleMemberAccessExpression(simpleNameOrMemberAccessExpression);
```
Comment (jmarolf) [inline src/Features/Core/Portable/GenerateMember/AbstractGenerateMemberService.cs]: >IsSimpleMemberAccessExpression [](start = 55, length = 30)

Why this name change? Was it always checking if it was a simple member access expression and the method was just poorly named?

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/CrefCompletionProviderTests.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/CrefCompletionProviderTests.cs]: I'm going to undo that.  But these tests are literally satan and need to go.

**Comment by CyrusNajmabadi** [comment]
> While a good change, shouldn't this go in a separate PR?

I don't like that philosophy.  That approach just means that general warts are not fixed up (because the cost of these tiny PRs is too high).  I'd much rather just integrate cleanup as i run into issues.

File: src/Workspaces/Core/Portable/Editing/SyntaxGenerator.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Editing/SyntaxGenerator.cs]: will do.

File: src/Features/Core/Portable/GenerateMember/AbstractGenerateMemberService.cs (hunk 3 above)
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/GenerateMember/AbstractGenerateMemberService.cs]: Yes.  It was poorly names.  I ended up adding IsSimpleMemberAccessExpression and then i realized it was the exact same as the existin gmethod.

**Comment by CyrusNajmabadi** [comment]
retest this please

**Comment by jmarolf** [comment]
> I don't like that philosophy. That approach just means that general warts are not fixed up (because the cost of these tiny PRs is too high). 

I hope the cost isn't too high.  I would like to set a good example for external contributors and have each PR have a central theme if possible.  I don't want people submitting PRs that are a grab bag of changes.  It should be simple enough to commit a change like adding a copyright header, cherry picking that over to a separate branch, and submitting a PR.  Otherwise we get PRs a lot of different changes and it can be hard to reason about.  When you removed `CrefCompletionSpeculatesOutsideTrivia` I couldn't see how it was necessary for this PR, but I thought "There must be something I'm missing then, and this PR must be more complex it initially looked".

File: src/Features/Core/Portable/UseObjectInitializer/AbstractUseObjectInitializerCodeFixProvider.cs [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+            var newRoot = editor.GetChangedRoot();
+            return document.WithSyntaxRoot(newRoot);
+        }
+
+        protected abstract TObjectCreationExpressionSyntax GetNewObjectCreation(
+            TObjectCreationExpressionSyntax objectCreation,
+            List<Match<TAssignmentStatementSyntax, TMemberAccessExpressionSyntax, TExpressionSyntax>> matches);
+
+        // protected abstract SyntaxNode AddTrivia(SyntaxNode syntaxNode);
```
Comment (jmarolf) [inline src/Features/Core/Portable/UseObjectInitializer/AbstractUseObjectInitializerCodeFixProvider.cs]: remove?

**Comment by CyrusNajmabadi** [comment]
> CrefCompletionSpeculatesOutsideTrivia 

That change was not intended to be part of this PR.  I accidentally pushed to the wrong branch.

**Comment by jmarolf** [comment]
@CyrusNajmabadi alright, no worries then ;)

**Comment by CyrusNajmabadi** [comment]
feel free to review that change here if you want: https://github.com/dotnet/roslyn/pull/13947  :)
